### PR TITLE
Minor: Remove unnecessary 'outputNode' on `SchemaKStream`.

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description 

Small clean up PR to remove the output node from `SchemaKStream`. This was previously needed as the physical building on the plan nodes resulted in a cloned and tweaked output node. With all the recent cleanup work this is no longer the case and so this can be cleaned up.

Along the way I: 
- slightly refactored `TransientQueryQueue` to allow me to remove the outputNode getter call
- Swapped limit from being `Optional<Integer>` to `OptionalInt`.
- Changed `JoinNode` to take explicit `DataSourceNode`s, rather than internally throw exceptions if anything else is passed. (Unrelated to main PR, but trivial).

### Testing done 

The usual.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: docs/changelog.rst [hunk 1]
```diff
@@ -8,6 +8,19 @@ KSQL 5.4.0 includes new features, including:
 
 * UDAFs support STRUCTs as parameters and return values.
 
+* KSQL now supports working with source data where the value is an anonymous Avro serialized
+  `ARRAY`, `MAP` or primitive type, for example `STRING` or `BIGINT`. Previously KSQL required all
+  Avro values to be a record.  For more information, refer to :ref:`ksql_single_field_wrapping`.
+
+* KSQL now supports working with source data where the value is an anonymous JSON serialized
```
Comment (JimGalasyn) [inline docs/changelog.rst]: Dupe paragraph?

File: docs/developer-guide/serialization.rst [hunk 2]
```diff
@@ -20,6 +22,148 @@ For more information on the formats that KSQL supports, see :ref:`ksql_formats`.
...
 
+.. _ksql_single_field_wrapping
+
+-------------------------
+Single field (un)wrapping
+-------------------------
+
+.. note:: The ``DELIMITED`` format is not effected by single field unwrapping.
+
+Controlling deserializing of single fields
```
Comment (JimGalasyn) [inline docs/developer-guide/serialization.rst]:
```suggestion
Control deserializing of single fields
```

File: docs/developer-guide/serialization.rst [hunk 3]
```diff
@@ -20,6 +22,148 @@ For more information on the formats that KSQL supports, see :ref:`ksql_formats`.
...
+==========================================
+
+When KSQL deserializes a Kafka message into a row, the key is deserialized into
+the key field and the message's value is deserialized into the value field(s).
```
Comment (JimGalasyn) [inline docs/developer-guide/serialization.rst]:
```suggestion
the key field, and the message's value is deserialized into the value field(s).
```

File: docs/developer-guide/serialization.rst [hunk 4]
```diff
@@ -20,6 +22,148 @@ For more information on the formats that KSQL supports, see :ref:`ksql_formats`.
...
+.. code:: json
+
+   {
+      "id": 134
+   }
+
+If your data contains only a single field, and that field is not wrapped within a JSON object,
+(or an Avro record if using the ``AVRO`` format), then you can use the ``WRAP_SINGLE_VALUE``
+property in the ``WITH`` clause of your :ref:`CREATE TABLE <create-table>` or
+:ref:`CREATE STREAM <create-stream>` statements. Setting the property to ``false`` will inform KSQL
```
Comment (JimGalasyn) [inline docs/developer-guide/serialization.rst]:
```suggestion
:ref:`CREATE STREAM <create-stream>` statements. Setting the property to ``false`` tells KSQL
```

File: docs/developer-guide/serialization.rst [hunk 5]
```diff
@@ -20,6 +22,148 @@ For more information on the formats that KSQL supports, see :ref:`ksql_formats`.
...
+that the value is not wrapped, i.e. the example above would be a JSON number:
```
Comment (JimGalasyn) [inline docs/developer-guide/serialization.rst]:
```suggestion
that the value is not wrapped, so the example above would be a JSON number:
```

File: docs/developer-guide/serialization.rst [hunk 6]
```diff
@@ -20,6 +22,148 @@ For more information on the formats that KSQL supports, see :ref:`ksql_formats`.
...
+.. code:: json
+
+     134
+
+For example, the following creates a table where the values in the underlying
+topic have been serialized as an anonymous JSON number:
+
+.. code:: sql
+
+    CREATE TABLE x (ID INT) WITH (WRAP_SINGLE_VALUE=false, ...);
```
Comment (JimGalasyn) [inline docs/developer-guide/serialization.rst]: Maybe a slightly more descriptive name than `x`. :)

File: docs/developer-guide/serialization.rst [hunk 7]
```diff
@@ -20,6 +22,148 @@ For more information on the formats that KSQL supports, see :ref:`ksql_formats`.
...
+
+
+If a statement does not explicitly set the value wrapping, the system
```
Comment (JimGalasyn) [inline docs/developer-guide/serialization.rst]:
```suggestion
If a statement doesn't set the value wrapping explicitly, the system
```

File: docs/developer-guide/serialization.rst [hunk 8]
```diff
@@ -20,6 +22,148 @@ For more information on the formats that KSQL supports, see :ref:`ksql_formats`.
...
+default, defined by ``ksql.persistence.wrap.single.values``, will be used.
```
Comment (JimGalasyn) [inline docs/developer-guide/serialization.rst]:
```suggestion
default, which is defined by ``ksql.persistence.wrap.single.values``.
```

File: docs/developer-guide/serialization.rst [hunk 9]
```diff
@@ -20,6 +22,148 @@ For more information on the formats that KSQL supports, see :ref:`ksql_formats`.
...
+The system default can be changed. For more information, see the
```
Comment (JimGalasyn) [inline docs/developer-guide/serialization.rst]:
```suggestion
You can change the system default. For more information, see
```

File: docs/developer-guide/serialization.rst [hunk 10]
```diff
@@ -20,6 +22,148 @@ For more information on the formats that KSQL supports, see :ref:`ksql_formats`.
...
+:ref:`ksql-persistence-wrap-single-values` documentation.
```
Comment (JimGalasyn) [inline docs/developer-guide/serialization.rst]:
```suggestion
:ref:`ksql-persistence-wrap-single-values`.
```

File: docs/developer-guide/serialization.rst [hunk 11]
```diff
@@ -20,6 +22,148 @@ For more information on the formats that KSQL supports, see :ref:`ksql_formats`.
...
+
+Controlling serialization of single fields
```
Comment (JimGalasyn) [inline docs/developer-guide/serialization.rst]:
```suggestion
Control serialization of single fields
```

File: docs/developer-guide/serialization.rst [hunk 12]
```diff
@@ -20,6 +22,148 @@ For more information on the formats that KSQL supports, see :ref:`ksql_formats`.
...
+==========================================
+
+When KSQL serializes a row into a Kafka message, the key field is serialized
+into the message's key, and any value field(s) are serialized into the
+message's value.
+
+By default, if the value has only a single field, KSQL will serialize the
```
Comment (JimGalasyn) [inline docs/developer-guide/serialization.rst]:
```suggestion
By default, if the value has only a single field, KSQL serializes the
```

File: docs/developer-guide/serialization.rst [hunk 13]
```diff
@@ -20,6 +22,148 @@ For more information on the formats that KSQL supports, see :ref:`ksql_formats`.
...
+single field as a named field within a record. However, this does not always match
```
Comment (JimGalasyn) [inline docs/developer-guide/serialization.rst]:
```suggestion
single field as a named field within a record. However, this doesn't always match
```

File: docs/developer-guide/serialization.rst [hunk 14]
```diff
@@ -20,6 +22,148 @@ For more information on the formats that KSQL supports, see :ref:`ksql_formats`.
...
+the requirements of downstream consumers. KSQL allows the value to also be serialized
```
Comment (JimGalasyn) [inline docs/developer-guide/serialization.rst]:
```suggestion
the requirements of downstream consumers, so KSQL allows the value to be serialized
```

File: docs/developer-guide/serialization.rst [hunk 15]
```diff
@@ -20,6 +22,148 @@ For more information on the formats that KSQL supports, see :ref:`ksql_formats`.
...
+
+.. code:: sql
+
+    CREATE STREAM x (f0 INT, f1 STRING) WITH (VALUE_FORMAT='JSON', ...);
+    CREATE STREAM y AS SELECT f0 FROM x;
+
+The second statement defines a stream with only a single field in the value,
+named ``f0``.
+
+When KSQL writes out the result to Kafka it will, by default, persist the
```
Comment (JimGalasyn) [inline docs/developer-guide/serialization.rst]:
```suggestion
By default, when KSQL writes out the result to Kafka, it persists the
```

File: docs/developer-guide/serialization.rst [hunk 16]
```diff
@@ -20,6 +22,148 @@ For more information on the formats that KSQL supports, see :ref:`ksql_formats`.
...
+single field as a named field within a JSON object, (or an Avro record if
+using the ``AVRO`` format):
```
Comment (JimGalasyn) [inline docs/developer-guide/serialization.rst]:
```suggestion
using the ``AVRO`` format:
```

File: docs/developer-guide/serialization.rst [hunk 17]
```diff
@@ -20,6 +22,148 @@ For more information on the formats that KSQL supports, see :ref:`ksql_formats`.
...
+.. code:: sql
+
+    CREATE STREAM x (f0 INT, f1 STRING) WITH (VALUE_FORMAT='JSON', ...);
+    CREATE STREAM y AS SELECT f0 FROM x;
+
+The second statement defines a stream with only a single field in the value,
+named ``f0``.
+
+When KSQL writes out the result to Kafka it will, by default, persist the
+single field as a named field within a JSON object, (or an Avro record if
```
Comment (JimGalasyn) [inline docs/developer-guide/serialization.rst]:
```suggestion
single field as a named field within a JSON object, or as an Avro record if
```

File: docs/developer-guide/serialization.rst [hunk 18]
```diff
@@ -20,6 +22,148 @@ For more information on the formats that KSQL supports, see :ref:`ksql_formats`.
...
+
+Then you can use the ``WRAP_SINGLE_VALUE`` property in your statement.
+
+For example,
+
+.. code:: sql
+
+    CREATE STREAM y WITH(WRAP_SINGLE_VALUE=false) AS SELECT f0 FROM x;
+
+If a statement does not explicitly set the value wrapping, the system
```
Comment (JimGalasyn) [inline docs/developer-guide/serialization.rst]:
```suggestion
If a statement doesn't explicitly set the value wrapping, the system
```

File: docs/developer-guide/serialization.rst [hunk 19]
```diff
@@ -20,6 +22,148 @@ For more information on the formats that KSQL supports, see :ref:`ksql_formats`.
...
+default, defined by ``ksql.persistence.wrap.single.values``, will be used.
```
Comment (JimGalasyn) [inline docs/developer-guide/serialization.rst]:
```suggestion
the system default, which is defined by ``ksql.persistence.wrap.single.values``.
```

File: docs/developer-guide/serialization.rst [hunk 20]
```diff
@@ -20,6 +22,148 @@ For more information on the formats that KSQL supports, see :ref:`ksql_formats`.
...
+The system default can be changed. For more information, see the
```
Comment (JimGalasyn) [inline docs/developer-guide/serialization.rst]:
```suggestion
You can change the system default. For more information, see
```

File: docs/developer-guide/serialization.rst [hunk 21]
```diff
@@ -20,6 +22,148 @@ For more information on the formats that KSQL supports, see :ref:`ksql_formats`.
...
+:ref:`ksql-persistence-wrap-single-values` documentation.
```
Comment (JimGalasyn) [inline docs/developer-guide/serialization.rst]:
```suggestion
:ref:`ksql-persistence-wrap-single-values`.
```

File: docs/developer-guide/serialization.rst [hunk 22]
```diff
@@ -20,6 +22,148 @@ For more information on the formats that KSQL supports, see :ref:`ksql_formats`.
...
+
+Examples
```
Comment (JimGalasyn) [inline docs/developer-guide/serialization.rst]:
```suggestion
Single-field serialization examples
```

File: docs/developer-guide/serialization.rst [hunk 23]
```diff
@@ -20,6 +22,148 @@ For more information on the formats that KSQL supports, see :ref:`ksql_formats`.
...
+========
```
Comment (JimGalasyn) [inline docs/developer-guide/serialization.rst]:
```suggestion
======================
```

File: docs/developer-guide/serialization.rst [hunk 24]
```diff
@@ -92,16 +236,44 @@ And a JSON value of:
...
 
 ----
 Avro
 ----
 
-The ``AVRO`` format supports Avro binary serialization of all of KSQL's ref:`data types <data-types>`.
-
-The format supports Avro records only. Top-level primitives, arrays and maps are not supported at
-this time.
+The ``AVRO`` format supports Avro binary serialization of all of KSQL's ref:`data types <data-types>`
```
Comment (JimGalasyn) [inline docs/developer-guide/serialization.rst]:
```suggestion
The ``AVRO`` format supports Avro binary serialization of all KSQL ref:`data types <data-types>`,
```

File: docs/developer-guide/serialization.rst [hunk 25]
```diff
@@ -92,16 +236,44 @@ And a JSON value of:
...
+including records and top-level primitives, arrays and maps.
```
Comment (JimGalasyn) [inline docs/developer-guide/serialization.rst]:
```suggestion
including records and top-level primitives, arrays, and maps.
```

File: docs/developer-guide/serialization.rst (hunk 2 above)
Comment (big-andy-coates) [reply docs/developer-guide/serialization.rst]: I think the original `Controlling` works better as the chapter is about 'controlling' deserialization of single fields...

File: docs/developer-guide/serialization.rst (hunk 6 above)
Comment (big-andy-coates) [reply docs/developer-guide/serialization.rst]: What would you prefer? 'Bob'?

File: docs/developer-guide/serialization.rst (hunk 11 above)
Comment (big-andy-coates) [reply docs/developer-guide/serialization.rst]: As above.

File: docs/developer-guide/serialization.rst (hunk 2 above)
Comment (JimGalasyn) [reply docs/developer-guide/serialization.rst]: Dat's cool, it's just that the two gerunds sound a bit awkward. :)

File: docs/developer-guide/serialization.rst (hunk 6 above)
Comment (JimGalasyn) [reply docs/developer-guide/serialization.rst]: Heh, I dunno, maybe something like `table-no-wrap-single-value` ... I'm all about verbose, self-describing object names, but ymmv.

**Pull request merged**